pkg/core/interop/contract/call.go

```diff
@@ -122,8 +122,44 @@ func callExFromNative(ic *interop.Context, caller util.Uint160, cs *state.Contra
 		initOff = md.Offset
 	}
 	ic.Invocations[cs.Hash]++
-	ic.VM.LoadNEFMethod(&cs.NEF, caller, cs.Hash, ic.VM.Context().GetCallFlags()&f,
-		hasReturn, methodOff, initOff, cb)
+	f = ic.VM.Context().GetCallFlags() & f
+
+	needWrap := f&(callflag.All^callflag.ReadOnly) != 0 // If the method is safe, then it's read-only and doesn't perform storage changes or emit notifications.
+	if !needWrap && ic.VM.Context().HasTryBlock() {     // If the method is not wrapped into try-catch block, then changes should be discarded anyway if exception occurs.
+		needWrap = true
+	}
+	if needWrap && ic.DAO != nil {
+		ic.DAO = ic.DAO.GetPrivate()
+	}
+	onUnload := func(commit bool, notificationsCount int) error {
+		if needWrap { // In case of CALL, CALLA, CALLL we don't need to commit/discard changes, unwrap DAO and change notificationsCount.
```

Comment:
CALL/CALLA/CALLL just don't exist here, every context they create must have `onUnload == nil`. And if we don't wrap, we don't need to set `onUnload` at all also.

Reply:
> CALL/CALLA/CALLL just don't exist here, every context they create must have onUnload == nil

That's true, the comment is removed.

>  And if we don't wrap, we don't need to set onUnload at all also.

Not exactly. Imagine if we skip wrapping because *safe* void method is called. We still have to push Null on context unloading, thus we have to set `onUnload`. But at the same time we don't need to unwrap because no wrapping was made and it's our tiny optimisation.

Reply:
OK, maybe `Null` problem justifies non-nil `onUnload` in some way, although I'm still not sure about it (let's see what the code looks like in the end).